### ISSUE TITLE
Need a complete example of a callable sort routine

### ISSUE THREAD

**Issue by tbrowder** [Raku/doc#3883]
The docs show an example, using infix, that a new user probably would not use.

A better example would be something that encapsulated this:

    sort({$^a.chars cmp $^b.chars}) # sorts by length of word, shortest first

I would provide the complete example, including how to use it for sorting a list of words, but I don't know how, hence this issue.

# UPDATE

Given a list of words, sort by word length:

    my @w = <abcde zero>
    say @w.sort;
    ["abcde", "zero"]
    @w .= sort({$^a.chars cmp $^b.chars});
    say @w.raku
    ["zero", "abcde"]

**Comment by Util** [comment]
I can provide any such example, but I am unclear on exactly what is being requested.

There are four classes whose .sort documentation appears on page https://docs.raku.org/routine/sort :
    Map, Supply, List, Any
and two classes mention `infix:` :
    List, Any
, yet I am still confused as to which class you are targeting, much less exactly which line, or the problem with the line.

I agree that the sort docs need improving, and hope to soon add a Sorting Tutorial page similar to https://docs.raku.org/language/grammar_tutorial .
In the meantime, would you clarify exactly which part of the sort docs this issue seeks to address?

**Comment by tbrowder** [comment]
Start with the Any class, but a List might make a better example. Use my sort requirement as I showed above, make it a "callable," and apply it to sorting a list of words.

**Comment by Util** [comment]
Hi, Tom!

(All written before your code update in the OP.)
I think I understand now; let me know if I am not spot-on.

Note: I changed `cmp` to `<=>`, as a better signal to a reader of the numeric nature of the comparison. I don't know if it is any more efficient, but why signal to a reader or to the compiler that we want Raku to figure out numeric-vs-alpha, when we can just say "numeric" explicitly?

Long version:

    # Sort comparator.
    # Words of the same length will remain in their original order.
    sub by_character_count ( $a, $b ) {
        return $a.chars <=> $b.chars;
    }
    my @play = <Let us impart what we have seen tonight unto young Hamlet>;
    say @play.sort(&by_character_count);
    # OUTPUT: «(us we Let what have seen unto young impart Hamlet tonight)␤»

Short version:

    sub by_character_count { $^a.chars <=> $^b.chars }
    say <Let us impart what we have seen tonight unto young Hamlet>.sort(&by_character_count);
    # OUTPUT: «(us we Let what have seen unto young impart Hamlet tonight)␤»

Are either of those close to what you had in mind?
If so, what tweaks are needed?
If not, please set me further straight on your intentions.

**Comment by tbrowder** [comment]
Pretty much spot on, Bruce. It wasn't clear to me what the callable's signature should be, nor how the signature is used (if at all) in the callable when used in sort(). Do we use '$^a' or '$a'?  From your example it seems it could just as well be '($b, $z)', though.

**Comment by jonathanstowe** [comment]
I thought that the signature _was_ considered if explicitly supplied, that is is the signature has one positional then the code is used for key extraction e.g.:

```
say <Let us impart what we have seen tonight unto young Hamlet>.sort(-> $c { $c.chars } );'
```

I'm pretty certain this is documented somewhere, but couldn't say where off the top of my head.

**Comment by JJ** [comment]
Maybe you can put this into a page and create a PR that solves this?

**Comment by raiph** [comment]
Sorting a list of pairs:
```
given (:1chris, :2chris, :2alice, :bob('03')) {

  # coerce values to strings
  say .sort: ~*.key;               # ((alice => 2 bob => 03 chris => 1 chris => 2)
  
  # coerce values to numbers
  say .sort: +*.value;             # (chris => 1 chris => 2 alice => 2 bob => 03)
  
  # .key within .value as number
  say .sort: { +.value, .key }     # (chris => 1 alice => 2 chris => 2 bob => 03)

  # .key within .value as string
  say .sort: { ~.value, .key }     # (bob => 03 chris => 1 alice => 2 chris => 2)

}
```

**Comment by Util** [comment]
@tbrowder and @jonathanstowe : 
Raku's sorting is conditioned on the `arity` of the code, no matter how that code is given. So, the signature of the comparator is critical.

This is mentioned here: https://docs.raku.org/type/List#routine_sort (Search for "and it accepts two arguments" and "accepts only one argument").

The details of how the 1-arity and 2-arity differ can *barely* be understood from the text+code, which is part of the reason I am writing a Tutorial/Fundamentals page.

The naming of $a as a variable in the parameter list vs $^a as a placeholder variable is a red herring here. They produce the same signature, as does an "arrow sub" version:

    raku -e '
        sub f1 ($a, $b)    { say $a }; 
        sub f2             { say $^a, $^b }; 
        my &f3 = -> $a, $b { say $a, $b }; 
        say "Arity={.arity}, Signature={.signature.raku}" for &f1, &f2, &f3;
    '
    Arity=2, Signature=:($a, $b)
    Arity=2, Signature=:($a, $b)
    Arity=2, Signature=:($a, $b)

In the PR I just submitted, I don't address any of this. I just added the minimal fix to the issue as it was raised, and will tackle the broader shortfall of sorting info in a separate ticket.

**Comment by Util** [comment]
> Maybe you can put this into a page and create a PR that solves this?

@JJ : done.
I have not merged the PR or closed the ticket, to give time for further discussion.
I will merge&close after 24 hours of no-comments, unless objections arise.
(I am resuming documentation work after a long absence, and trying to tread softly)

**Comment by JJ** [comment]
You are very welcome back!

**Comment by raiph** [comment]
Won't have time to look tonight, but just spotted Util on the scene before I quit for the night. <3 \o/
